src/groth16/circuit/builder/mod.rs

```diff
@@ -755,9 +913,333 @@ where
     ////////////////////////////////////////////////////////////////////////////////
 
-    // fn new_4_bit_compare(&mut self, lhs: [WireId; 4], rhs: [WireId; 4], cmp: [WireId; 3]) -> [WireId; 3] {
+    /// Requires that both the left and right inputs are either 0 or 1
+    fn new_less_than(&mut self, left: WireId, right: WireId) -> WireId {
+        let left_not = self.new_not(left);
+        self.new_and(left_not, right)
+    }
+
+    /// Requires that both the left and right inputs are either 0 or 1
+    fn new_greater_than(&mut self, left: WireId, right: WireId) -> WireId {
+        let right_not = self.new_not(right);
+        self.new_and(left, right_not)
+    }
+
+    /// Requires that both the left and right inputs are either 0 or 1
+    /// TODO: replace this with new_equal, need to define an Binary
+    /// WireId
+    fn new_equality(&mut self, left: WireId, right: WireId) -> WireId {
+        self.new_xnor(left, right)
+    }
+
+    /// ```
+    /// use zksnark::groth16::circuit::{Circuit, BinaryInput};
+    /// use zksnark::field::z251::Z251;
+    ///
+    /// let mut circuit = Circuit::<Z251>::new();
+    ///
+    /// let input_wire = circuit.new_word8();
+    /// let num = circuit.const_word8(5);
+    /// let eq =
+    ///     circuit.is_equal(&input_wire, &num);
+    ///
+    /// circuit.set_word8(&input_wire, 5);
+    /// assert_eq!(circuit.evaluate(eq), Z251::from(1));
+    ///
+    /// circuit.reset();
+    /// circuit.set_word8(&input_wire, 4);
+    /// assert_eq!(circuit.evaluate(eq), Z251::from(0));
+    ///
+    /// // Lets use is_equal with Word64 as well
+    /// //
+    /// // no need to reset() since I'm not modifying the inputs to
+    /// // the previous num, just shadowing it.
+    ///
+    /// let input_wire = circuit.new_word64();
+    /// let num = circuit.const_word64(1119784);
+    /// let eq =
+    ///     circuit.is_equal(&input_wire, &num);
+    ///
+    /// circuit.set_word64(&input_wire, 1119784);
+    /// assert_eq!(circuit.evaluate(eq), Z251::from(1));
+    ///
+    /// circuit.reset();
+    /// circuit.set_word64(&input_wire, 4);
+    /// assert_eq!(circuit.evaluate(eq), Z251::from(0));
+    ///
+    /// ```
+    pub fn is_equal<'a, Z, P>(&mut self, left: Z, right: P) -> WireId
+    where
+        Z: IntoIterator<Item = &'a WireId> + BinaryInput,
+        P: IntoIterator<Item = &'a WireId> + BinaryInput,
+    {
+        let mut l_iter = left.into_iter();
+        let mut r_iter = right.into_iter();
+
+        let base_case: WireId = self.new_equality(
+            *l_iter.next().expect("new_eqz: left input empty"),
+            *r_iter.next().expect("new_eqz: right input empty"),
+        );
+
+        l_iter
+            .zip_longest(r_iter)
+            .fold(base_case, |acc, x| match x {
+                Both(l, r) => {
+                    let eq = self.new_equality(*l, *r);
+                    self.new_and(eq, acc)
+                }
+                Left(l) => {
```

Comment:
I think it's a little strange that the equality circuit will fill in zeros when the inputs are different lengths. It makes more sense to me to either panic or maybe return false, but I think this circuit should probably assume that the inputs are equal length and panic when they are not.

Reply:
I agree, panicking is the best approach here.

Reply:
Sure, there may be times where a user would be working with different bus widths and then want to do some sort of comparison on them, but for that case I think it would be better to have a specific `padded_equal`, probably built on top of a `pad` function that could either pad a single input to a desired width or given two inputs return both of them where the shorter one has been padded to be as long as the other. I think this makes sense because these primitives that we are writing now are at the low level of functionality (even though we plan to use just a comparison in our larger system), and then convenience functions should be built on top. At this level I think the analogue could be assembly; sure it's not usually convenient to use, but if someone is writing assembly then they usually have a good reason and want full control and are willing to sacrifice some convenience.

Reply:
I prefer the proposal to force both left and right inputs to be the same type. Overall, I think it’s more likely that the programmer will erroneously compare different types that need to compare W8 and W64. If we encounter such a situation, then we can create a padding function for conversion, as Ross suggested.

Reply:
I like the proposal to enforce left and right inputs to be the same type. Implicit type conversion is typically removed in languages that want to encourage correctness, and when building a zkSNARK we definitely want that. 

I think, while it’s possible we would want to compare W8 and W64, it’s far more likely that we would do this erroneously (in all our proofs we need in the near future, all our data is a W64 except for hashes which aren’t used in comparisons with other non-hashes). To allow comparisons between different types we can add the pad function if we need it.

Reply:
You are right, I should no implicitly "convert" the types to compare them. I'll switch it back to:

```
    pub fn is_equal<'a, Z, P>(&mut self, left: Z, right: Z) -> WireId
    where
        Z: IntoIterator<Item = &'a WireId> + BinaryInput,
```

Reply:
That and all the other comparison functions